### DISCUSSION TITLE
How to implement Auth0 with Supabase within NextJS app router app?

### DISCUSSION THREAD

**Discussion by gshortdev** [supabase/supabase#30042] [Questions]
There doesn't seem to be a lot of documentation yet around how best to use Auth0 as a third party provider when using Supabase SSR library. I am able to get it to kind of work with standard Supabase JS library, but there has to be a better way. With @supabase/supabase-js I am able to pass the Auth0 id token to supabase. Auth0 works fine and is protecting the routes properly. But I want to be able to use RLS on my database queries from Supabase so I want to leverage Auth0 with Supabase Auth. Here is the code that kind of works (although it doesn't handle token refresh).

```
//This works fine in a server component
import { createClient } from '@supabase/supabase-js'
import { getSession } from "@auth0/nextjs-auth0";

async function getCategories() {

    const supabase = await createClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL,
        process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY,
        {
            accessToken: async () => {
                const session = await getSession()
                return session?.idToken ?? ""
            }
        }
    )

    return supabase.from('categories').select('id,name,description')
}
```
I haven't had luck getting this to work with the CreateServerClient in the SSR library though. Are there any tips or documentation anyone can point me at to handle this situation with the App router and SSR package?

I've tried to work off of these documents: 
[https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=app](https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=app) - Only covers Supabase Auth
[https://supabase.com/partners/integrations/auth0](https://supabase.com/partners/integrations/auth0) - only covers Pages router
[https://developer.auth0.com/resources/guides/web-app/nextjs/basic-authentication](https://developer.auth0.com/resources/guides/web-app/nextjs/basic-authentication) - Only covers Auth0 with app router

**Comment by harsh7800** [comment]
this is how you do it with ssr package
`export async function createSupabaseServerClient() {
  const cookieStore = cookies();

  return createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        getAll() {
          return cookieStore.getAll();
        },
        setAll(cookiesToSet) {
          try {
            cookiesToSet.forEach(({ name, value, options }) => {
              cookieStore.set(name, value, options);
            });
          } catch (error) {
            // The `set` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
      },
    }
  );
}`


`import { createServerClient } from "@supabase/ssr";
import { type NextRequest, NextResponse } from "next/server";

export const updateSession = async (request: NextRequest) => {
  try {
    // Create an unmodified response
    let response = NextResponse.next({
      request: {
        headers: request.headers,
      },
    });

    const supabase = createServerClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
      {
        cookies: {
          getAll() {
            return request.cookies.getAll();
          },
          setAll(cookiesToSet) {
            cookiesToSet.forEach(({ name, value }) =>
              request.cookies.set(name, value)
            );
            response = NextResponse.next({
              request,
            });
            cookiesToSet.forEach(({ name, value, options }) =>
              response.cookies.set(name, value, options)
            );
          },
        },
      }
    );

    // This will refresh session if expired - required for Server Components
    const {
      data: { user },
      error,
    } = await supabase.auth.getUser();

    // Protected routes
    const protectedRoutes = ["/marketplace", "/stake"];
    const isProtectedRoute = protectedRoutes.some((route) =>
      request.nextUrl.pathname.startsWith(route)
    );

    // If user is not logged in and trying to access a protected route, redirect to /sign-in
    if (isProtectedRoute && (error || !user)) {
      return NextResponse.redirect(new URL("/auth", request.url));
    }

    // // If the user is already logged in and trying to access the root ("/"), redirect to /marketplace
    // if (request.nextUrl.pathname === "/" && user) {
    //   return NextResponse.redirect(new URL("/marketplace", request.url));
    // }

    return response;
  } catch (e) {
    // If Supabase client could not be created, return unmodified response
    return NextResponse.next({
      request: {
        headers: request.headers,
      },
    });
  }
};
`
create a middleware an run the above function 

`import { type NextRequest } from "next/server";
import { updateSession } from "@/utils/supabase/middleware";

export async function middleware(request: NextRequest) {
  return await updateSession(request);
}

export const config = {
  matcher: [
    /*
     * Match all request paths except:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     * - images - .svg, .png, .jpg, .jpeg, .gif, .webp
     * Feel free to modify this pattern to include more paths.
     */
    "/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)",
  ],
};
`

**Reply by geeshort** [reply]
Thanks for the response! Where/how do I pass the Auth0 token in that flow?
